### PR TITLE
feat: Update backend to support Assessment Creation

### PR DESCRIPTION
**Pull request by brian386**
## Notion ticket link

[Update backend to support Assessment Creation](https://www.notion.so/uwblueprintexecs/Update-backend-to-support-Assessment-Creation-223d6e0d8100415a909f4d966bc31b54)

## Implementation description
*

## Steps to test
1.

## What should reviewers focus on?
* 

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/services/interfaces/testService.ts [hunk 1]
```diff
@@ -7,21 +7,27 @@ import {
...
   name: string;
-  /** the duration of the test */
-  duration: number;
   /** the UserDTO for the admin */
   admin: UserDTO;
   /** an array of questions on the test */
   questions: QuestionComponent[][];
   /** the grade of the student */
   grade: number;
+  /** the type of assessment */
```
Comment (brian386) [inline backend/services/interfaces/testService.ts:25]: I didn't actually know what the fields mean so these comments might not be that descriptive

File: backend/graphql/types/testType.ts [hunk 2]
```diff
@@ -10,6 +10,18 @@ const testType = gql`
     SHORT_ANSWER
   }
 
+  enum AssessmentType {
+    Beginning
+    End
```
Comment (joyce-shi) [inline backend/graphql/types/testType.ts]: nit: can we make the values all caps to be consistent with the `QuestionComponentTypeEnum`? i.e `Beginning` -> `BEGINNING`

File: backend/graphql/types/testType.ts [hunk 3]
```diff
@@ -10,6 +10,18 @@ const testType = gql`
...
+  }
+
+  enum Status {
+    Draft
+    Published
+    Archived
+    Deleted
```
Comment (joyce-shi) [inline backend/graphql/types/testType.ts]: same as above! ^

File: backend/services/interfaces/testService.ts [hunk 4]
```diff
@@ -7,21 +7,27 @@ import {
...
+  assessmentType: AssessmentType;
+  /** the status of the assessment * */
+  status: AssessmentStatus;
+  /** the country of the test* */
+  curriculumCountry: string;
+  /** the region of the test */
```
Comment (joyce-shi) [inline backend/services/interfaces/testService.ts]:
```suggestion
  /** the region that the test is to be administered in*/
```

File: backend/services/interfaces/testService.ts [hunk 5]
```diff
@@ -7,21 +7,27 @@ import {
...
   admin: UserDTO;
   /** an array of questions on the test */
   questions: QuestionComponent[][];
   /** the grade of the student */
   grade: number;
+  /** the type of assessment */
+  assessmentType: AssessmentType;
+  /** the status of the assessment * */
+  status: AssessmentStatus;
+  /** the country of the test* */
```
Comment (joyce-shi) [inline backend/services/interfaces/testService.ts]:
```suggestion
  /**the country that the test is to be administered in * */
```

File: backend/models/test.model.ts [hunk 6]
```diff
@@ -123,8 +125,26 @@ const TestSchema: Schema = new Schema(
...
+    assessmentType: {
+      type: String,
+      required: true,
+      enum: ["Beginning", "End"],
+    },
+    status: {
+      type: String,
+      required: true,
+      enum: ["Draft", "Published", "Archived", "Deleted"],
+    },
```
Comment (joyce-shi) [inline backend/models/test.model.ts:155]: I think we can define both `assessmentType` and `status` enums above! And then no need to define a separate type. i.e line 54 in `backend/types.ts`. For an example of what I mean, see [definition](https://github.com/uwblueprint/jump-math/blob/bc6ee0458db1270225a12646af94cfe290435a97/backend/models/testSession.model.ts#L6) of `GradingStatus`!

**Review by joyce-shi — APPROVED** [review]
LGTM! Left some small comments.

File: backend/graphql/types/testType.ts [hunk 7]
```diff
@@ -88,18 +100,24 @@ const testType = gql`
...
     id: ID!
     name: String!
-    duration: Int!
     admin: UserDTO!
     questions: [[QuestionComponent]]!
     grade: Int!
+    assessmentType: AssessmentType!
+    curriculumCountry: String!
+    curriculumRegion: String!
+    Status: Status!
```
Comment (carissa-tang) [inline backend/graphql/types/testType.ts]:
```suggestion
    assessmentType: AssessmentTypeEnum!
    curriculumCountry: String!
    curriculumRegion: String!
    Status: StatusEnum!
```
can we rename these for consistency with other files?

File: backend/graphql/types/testType.ts [hunk 8]
```diff
@@ -109,7 +127,7 @@ const testType = gql`
   extend type Mutation {
     createTest(test: TestRequestDTO!): TestResponseDTO!
     updateTest(id: ID!, test: TestRequestDTO!): TestResponseDTO!
-    deleteTestById(id: ID!): ID
+    deleteTestById(id: ID!): ID!
```
Comment (carissa-tang) [inline backend/graphql/types/testType.ts]: curious to know if this was added because you ran into an error? but i do think it should have one

if you ran into an error, we might want to fix any delete functions with a missing `!` in the other files

File: backend/graphql/types/testType.ts (hunk 2 above)
Comment (brian386) [reply backend/graphql/types/testType.ts]: There are places in the frontend that use Beginning and End. Should I change those as well?

Comment (joyce-shi) [reply backend/graphql/types/testType.ts]: yes please! thanks 😄

File: backend/graphql/types/testType.ts (hunk 8 above)
Comment (brian386) [reply backend/graphql/types/testType.ts]: I don't think it got errors. I was just removing `!` to troubleshoot another error (because the existing tests have null for the added fields and was erroring and I fixed it by changing the data in the database). I guess when I added them back I accidentally added an extra one

File: backend/graphql/index.ts [hunk 9]
```diff
@@ -61,7 +61,7 @@ const graphQLMiddlewares = {
     simpleEntities: authorizedByAllRoles(),
     userById: authorizedByAdmin(),
     users: authorizedByAdmin(),
-    tests: authorizedByAllRoles(),
+    // tests: authorizedByAllRoles(),
```
Comment (carissa-tang) [inline backend/graphql/index.ts]: can you uncomment this?

File: backend/graphql/types/testType.ts (hunk 8 above)
Comment (carissa-tang) [reply backend/graphql/types/testType.ts]: wait so there shouldn't be an extra `!`? (remove it pls if not)

File: backend/models/test.model.ts [hunk 10]
```diff
@@ -98,6 +106,14 @@ export interface Test extends Document {
   questions: QuestionComponent[][];
   /** The intended grade the test was made for */
   grade: number;
+  /** the type of assessment */
```
Comment (carissa-tang) [inline backend/models/test.model.ts:109]: can you match the documentation in `backend/services/interfaces/testService.ts` with this file

File: backend/services/interfaces/testService.ts [hunk 11]
```diff
@@ -44,14 +52,20 @@ export type QuestionComponentRequest = Omit<QuestionComponent, "metadata"> &
...
   name: string;
-  /** the duration of the test */
-  duration: number;
   /** the UserDTO for the admin */
   admin: string;
   /** an ordered array of questions on the test */
   questions: QuestionComponentRequest[][];
   /** the grade of the student */
   grade: number;
+  /** the type of assessment */
```
Comment (carissa-tang) [inline backend/services/interfaces/testService.ts:61]: match the documentation with the above ^ (for TestResponseDTO)
